### PR TITLE
[Kernel] use flashinfer for gdn prefill

### PR DESCRIPTION
**Pull request by ZJY0516**
## Purpose
flashinfer introduced prefill gdn kernel in https://github.com/flashinfer-ai/flashinfer/pull/2276

## Test Plan
```
vllm bench serve --model Qwen/Qwen3-Next-80B-A3B-Instruct --dataset-name random \
                                                               --num-prompts 512 \
                                                               --random-input-len 4096 \
                                                               --random-output-len 512
```
## Test Result
main
```
============ Serving Benchmark Result ============
Successful requests:                     512       
Failed requests:                         0         
Benchmark duration (s):                  107.53    
Total input tokens:                      2097152   
Total generated tokens:                  262144    
Request throughput (req/s):              4.76      
Output token throughput (tok/s):         2437.78   
Peak output token throughput (tok/s):    7680.00   
Peak concurrent requests:                512.00    
Total token throughput (tok/s):          21940.02  
---------------Time to First Token----------------
Mean TTFT (ms):                          42234.71  
Median TTFT (ms):                        40749.87  
P99 TTFT (ms):                           78822.57  
-----Time per Output Token (excl. 1st token)------
Mean TPOT (ms):                          119.60    
Median TPOT (ms):                        123.75    
P99 TPOT (ms):                           167.04    
---------------Inter-token Latency----------------
Mean ITL (ms):                           119.60    
Median ITL (ms):                         69.34     
P99 ITL (ms):                            318.22    
==================================================
```

this pr
```
============ Serving Benchmark Result ============
Successful requests:                     512       
Failed requests:                         0         
Benchmark duration (s):                  100.25    
Total input tokens:                      2097152   
Total generated tokens:                  262144    
Request throughput (req/s):              5.11      
Output token throughput (tok/s):         2615.02   
Peak output token throughput (tok/s):    7936.00   
Peak concurrent requests:                512.00    
Total token throughput (tok/s):          23535.15  
---------------Time to First Token----------------
Mean TTFT (ms):                          34522.29  
Median TTFT (ms):                        33466.90  
P99 TTFT (ms):                           71442.06  
-----Time per Output Token (excl. 1st token)------
Mean TPOT (ms):                          120.45    
Median TPOT (ms):                        123.74    
P99 TPOT (ms):                           172.21    
---------------Inter-token Latency----------------
Mean ITL (ms):                           120.45    
Median ITL (ms):                         69.12     
P99 ITL (ms):                            326.51    
==================================================
```

## Accuracy Test

```
vllm serve Qwen/Qwen3-Next-80B-A3B-Instruct -tp 4
lm_eval --model local-chat-completions --model_args model=Qwen/Qwen3-Next-80B-A3B-Instruct,base_url=http://localhost:8000/v1/chat/completions,num_concurrent=280 --tasks gsm8k --apply_chat_template --num_fewshot 5 --trust_remote_code
```
main
|Tasks|Version|     Filter     |n-shot|  Metric   |   |Value |   |Stderr|
|-----|------:|----------------|-----:|-----------|---|-----:|---|-----:|
|gsm8k|      3|flexible-extract|     5|exact_match|↑  |0.8484|±  |0.0099|
|     |       |strict-match    |     5|exact_match|↑  |0.6376|±  |0.0132|

This pr
|Tasks|Version|     Filter     |n-shot|  Metric   |   |Value |   |Stderr|
|-----|------:|----------------|-----:|-----------|---|-----:|---|-----:|
|gsm8k|      3|flexible-extract|     5|exact_match|↑  |0.8453|±  |0.0100|
|     |       |strict-match    |     5|exact_match|↑  |0.6376|±  |0.0132|
---
<details>
<summary> Essential Elements of an Effective PR Description Checklist </summary>

- [ ] The purpose of the PR, such as "Fix some issue (link existing issues this PR will resolve)".
- [ ] The test plan, such as providing test command.
- [ ] The test results, such as pasting the results comparison before and after, or e2e results
- [ ] (Optional) The necessary documentation update, such as updating `supported_models.md` and `examples` for a new model.
- [ ] (Optional) Release notes update. If your change is user facing, please update the release notes draft in the [Google Doc](https://docs.google.com/document/d/1YyVqrgX4gHTtrstbq8oWUImOyPCKSGnJ7xtTpmXzlRs/edit?tab=t.0).
</details>

### CODE REVIEW

File: vllm/model_executor/models/qwen3_next.py [hunk 1]
```diff
@@ -7,6 +7,7 @@
 
 import torch
 from einops import rearrange
+from flashinfer.gdn_prefill import chunk_gated_delta_rule as chunk_gated_delta_rule_fi
```
Comment (mgoin) [inline vllm/model_executor/models/qwen3_next.py]: This should be lazy imported right before the function call since flashinfer is only installed for cuda platforms

File: vllm/model_executor/models/qwen3_next.py [hunk 2]
```diff
@@ -99,6 +103,63 @@
...
+    beta,
+    initial_state,
+    output_final_state,
+    cu_seqlens,
+    head_first=False,
+    use_qk_l2norm_in_kernel=True,
+):
+    if (
+        current_platform.is_cuda()
+        and current_platform.get_device_capability().major == 9
```
Comment (mgoin) [inline vllm/model_executor/models/qwen3_next.py]: nit: use current_platform.is_device_capability(90)

File: vllm/model_executor/models/qwen3_next.py [hunk 3]
```diff
@@ -99,6 +103,63 @@
 KVCache = tuple[torch.Tensor, torch.Tensor]
 
 
+def chunk_gated_delta_rule(
```
Comment (vadiklyutiy) [inline vllm/model_executor/models/qwen3_next.py]: I think it is better to wrap the logic of chosen of kernel to class (inherited from `CustomOp`) and place it to separate file. 
If do so we can reuse the code of GDN for another models.

Comment (ZJY0516) [reply vllm/model_executor/models/qwen3_next.py]: CustomOp are used to dispatch according to platform. I'm not sure whether we could use this here

Comment (vadiklyutiy) [reply vllm/model_executor/models/qwen3_next.py]: FlashInfer kernel is for Cuda only

**Comment by mayuyuace** [comment]
TPOT is slower with this PR. 
And may I know did you test the main branch at first running?
You should test second running time with main branch, because triton path uses auto tunning at first running, what causes prefill stage is slower about tenfold.
If performance data you give is from first running, auto tunning of triton is the main reason why the throughput of this PR is better than triton path instead of kernel performance.

**Comment by vadiklyutiy** [comment]
@ZJY0516  #33291 was merged
It might solve accuracy issue

File: vllm/model_executor/models/qwen3_next.py (hunk 3 above)
Comment (ZJY0516) [reply vllm/model_executor/models/qwen3_next.py]: > FlashInfer kernel is for Cuda only

Yes. I mean customop is used to dispatch kernels for different hardware. What we need here is more like attention backend. I would suggest leave it for following pr

File: vllm/model_executor/models/qwen3_next.py [hunk 4]
```diff
@@ -101,6 +104,64 @@
...
+        g = rearrange(g, "1 b h -> b h").contiguous()
+        beta = rearrange(beta, "1 b h -> b h").contiguous()
+        fi_state = initial_state.to(torch.float32)
+        fi_g = g.to(torch.float32)
+        fi_beta = beta.to(torch.float32)
+        return chunk_gated_delta_rule_fi(
+            q=q,
+            k=k,
+            v=v,
+            g=torch.exp(fi_g),
```
Comment (vadiklyutiy) [inline vllm/model_executor/models/qwen3_next.py]: as I understand FI's GDN fused more ops than current GDA and here we revert previous `ln`.
I think we should we should use new FI's GDN in full power and avoid sequential `exp(ln(x))`

File: vllm/model_executor/models/qwen3_next.py (hunk 3 above)
Comment (vadiklyutiy) [reply vllm/model_executor/models/qwen3_next.py]: The main point I wanted to make is that I think it’s worth making this more universal. The current implementation works only for Qwen3-Next, but GDN is used in the Kimi model as well, and it’s likely that new models will use GDN in the future. It would be great to make these changes in a way that enables automatic support across all models without requiring any model-specific modifications.

**Review by vadiklyutiy — CHANGES_REQUESTED** [review]

File: vllm/model_executor/models/qwen3_next.py (hunk 4 above)
Comment (ZJY0516) [reply vllm/model_executor/models/qwen3_next.py]: Actually, we don't have previous `ln`. 

$g = -\exp(A_{\log}) \cdot \text{softplus}(a + \mathrm{dt\_bias})$.

FI GDN expects exponential input and then applies the logarithm internally. These steps are not included in the Triton kernel.

**Comment by ZJY0516** [comment]
And for kimi-linear, though it uses gdn attention backend, but it actually uses another attention: kda, not as same as gdn. It means it doesn't use `chunk_gated_delta_rule` @vadiklyutiy

File: vllm/model_executor/models/qwen3_next.py (hunk 4 above)
Comment (vadiklyutiy) [reply vllm/model_executor/models/qwen3_next.py]: ok, got it

File: vllm/model_executor/models/qwen3_next.py [hunk 5]
```diff
@@ -101,6 +104,59 @@
...
+class ChunkGatedDeltaRule(nn.Module):
+    def __init__(self) -> None:
+        super().__init__()
+        if current_platform.is_cuda() and current_platform.is_device_capability(90):
+            logger.info_once(
+                "Using FlashInfer GDN prefill kernel on CUDA compute capability 9.x"
+            )
+            self.forward = fi_chunk_gated_delta_rule
+        else:
+            self.forward = fla_chunk_gated_delta_rule
```
Comment (youkaichao) [inline vllm/model_executor/models/qwen3_next.py]: directly changing `self.forward` looks hacky.

you can have something like `self.func = fi_chunk_gated_delta_rule`, and call it inside `def forward()` .

I think making it a `CustomOp` also makes sense. Other hardware platforms might need it in the future.

**Review by youkaichao — COMMENTED** [review]
LGTM in general

File: vllm/model_executor/models/qwen3_next.py (hunk 5 above)
Comment (ZJY0516) [reply vllm/model_executor/models/qwen3_next.py]: Done. I don't use Customop here because we need to dispatch according to compute capability

File: vllm/model_executor/models/qwen3_next.py [hunk 6]
```diff
@@ -101,6 +105,112 @@
...
+
+
+@CustomOp.register("chunk_gated_delta_rule")
+class ChunkGatedDeltaRule(CustomOp):
+    def __init__(self) -> None:
+        super().__init__()
+        if current_platform.is_cuda() and current_platform.is_device_capability(90):
+            logger.info_once(
+                "Using FlashInfer GDN prefill kernel on CUDA compute capability 90"
+            )
```
Comment (youkaichao) [inline vllm/model_executor/models/qwen3_next.py:157]: check flashinfer version and confirm if it has this function?

**Review by youkaichao — APPROVED** [review]
the direction looks good to me. @mgoin might also have some feedback

File: vllm/model_executor/models/qwen3_next.py (hunk 6 above)
Comment (ZJY0516) [reply vllm/model_executor/models/qwen3_next.py:157]: Now vllm uses flashinfer==0.6.3 and it should has this kernel(introduced in 0.6.0). Do I need to consider earlier version?

Comment (youkaichao) [reply vllm/model_executor/models/qwen3_next.py:157]: i think it's better to have some attribute check.

Comment (mgoin) [reply vllm/model_executor/models/qwen3_next.py:157]: We do attribute checks sometimes, but honestly I think it isn't needed. We will always move forward in versions and just need a test to make sure functions we depend on don't break.
We should use the has_flashinfer() check though

File: vllm/model_executor/models/qwen3_next.py [hunk 7]
```diff
@@ -101,6 +105,112 @@
...
+
+    # use flashinfer implementation
+    q = rearrange(q, "1 l h d -> l h d").contiguous()
+    k = rearrange(k, "1 l h d -> l h d").contiguous()
+    v = rearrange(v, "1 l h d -> l h d").contiguous()
+    g = rearrange(g, "1 b h -> b h").contiguous()
+    beta = rearrange(beta, "1 b h -> b h").contiguous()
+    fi_state = initial_state.to(torch.float32)
+    fi_g = g.to(torch.float32)
+    fi_beta = beta.to(torch.float32)
```
Comment (mgoin) [inline vllm/model_executor/models/qwen3_next.py]: nit: These could be slow since they will be wrapped in a custom op and unavailable to torch.compile to fuse. For the `rearrange(q, "1 l h d -> l h d").contiguous()`, why isn't this just `q.unsqueeze()`?

File: vllm/model_executor/models/qwen3_next.py [hunk 8]
```diff
@@ -101,6 +105,112 @@
...
+        super().__init__()
+        if current_platform.is_cuda() and current_platform.is_device_capability(90):
+            logger.info_once(
+                "Using FlashInfer GDN prefill kernel on CUDA compute capability 90"
+            )
+            self._forward_method = self.forward_cuda
+        else:
+            self._forward_method = self.forward_native
+
+    def forward_cuda(
```
Comment (mgoin) [inline vllm/model_executor/models/qwen3_next.py:162]: It might be confusing for future users to have this forward_cuda function only work on SM90 but no asserts within it about the CC

File: vllm/model_executor/models/qwen3_next.py [hunk 9]
```diff
@@ -101,6 +105,112 @@
...
+    q,
+    k,
+    v,
+    g,
+    beta,
+    initial_state,
+    output_final_state,
+    cu_seqlens,
+    head_first=False,
+    use_qk_l2norm_in_kernel=True,
```
Comment (mgoin) [inline vllm/model_executor/models/qwen3_next.py]: missing type annotation

File: vllm/model_executor/models/qwen3_next.py (hunk 8 above)
Comment (ZJY0516) [reply vllm/model_executor/models/qwen3_next.py:162]: But if we assert in every forward, it will hurt performance.

File: vllm/model_executor/models/qwen3_next.py (hunk 7 above)
Comment (ZJY0516) [reply vllm/model_executor/models/qwen3_next.py]: Yes, this is the limitation of linear attention now. Because we cannot split one core attention op like full attention cleanly

**Comment by vadiklyutiy** [comment]
I still think we should move GDN function in separate file. Right now qwen35 just import from Qwen3-next them. It looks a bit awkwardly. But this is don't block this PR and we can do it later. 
BTW to double check qwen3.5 will use FI now? Or we should adopt qwen35 code as well?

**Review by mgoin — APPROVED** [review]
Thanks!
